bin/wasm-node/rust/src/sync_service.rs

```diff
@@ -515,6 +530,27 @@ async fn start_sync(
                         sync = sync_idle.into();
                     }
                 },
+
+                (request_id, result) = pending_grandpa_requests.select_next_some() => {
+                    pending_requests.remove(&request_id);
+
+                    // A GrandPa warp sync request has been finished.
+                    // `result` is an error if the block request got cancelled by the sync state
+                    // machine.
+                    if let Ok(result) = result {
+                        // Inject the result of the request into the sync state machine.
+                        let new_sync = sync_idle.grandpa_warp_sync_response(
+                            request_id,
+                            result.ok(),
+                        );
+
+                        sync = new_sync;
```

Comment:
I think we want to handle `Err(..)` warp sync responses somewhere, either here or in `sync/all`. Do you have a preference on where?

Reply:
Errors should directly be handled in the GrandPa warp sync state machine. If the verification fails, the GrandPa warp sync code should switch back to `WaitingForSources` or `WarpSyncRequest` to try do another request.

Reply:
I mean we already do that with the `Option` that we pass in, but I'm talking more about disconnecting or giving a bad reputation to nodes that give out bad responses. That seems out of scope for grandpa warp sync.

Reply:
> I mean we already do that with the Option that we pass in,

The `Option` covers the situation where someone sends us random garbage data. But if someone sends us for example an incomplete list of signatures, we will jump to the `Verifier` state and fail. In that situation, we should jump back to `WaitingForSources` or `WarpSyncRequest`, and not just return an error and kill the entire GrandPa warp sync state machine.

> but I'm talking more about disconnecting or giving a bad reputation to nodes that give out bad responses. That seems out of scope for grandpa warp sync.

I don't think we need to disconnect or maintain reputations, but for what it's worth the `grandpa_warp_sync_response` function will eventually return an enum indicating what happened after we have received the response (and which will notably contain a follow-up request if relevant).

Reply:
> the `grandpa_warp_sync_response` function will eventually return an enum indicating what happened after we have received the response (and which will notably contain a follow-up request if relevant).

Ah, that's a good idea!